vanguard.go

```diff
@@ -136,33 +142,10 @@ func NewTranscoder(services []*Service, opts ...TranscoderOption) (*Transcoder,
 		if err := transcoder.registerService(svc, resolvedOpts); err != nil {
 			return nil, err
 		}
-		if len(resolvedOpts.protocols) == 1 {
-			_, ok := resolvedOpts.protocols[ProtocolREST]
-			if ok {
-				restOnlyServices = append(restOnlyServices, svc.schema)
-			}
-		}
 	}
 	if err := transcoder.registerRules(transcoderOpts.rules); err != nil {
 		return nil, err
 	}
-
-	// Finally, check that any services with only REST as target protocol
-	// actually have at least one method with REST mappings.
-	for _, svcDesc := range restOnlyServices {
-		methods := svcDesc.Methods()
-		var numSupportedMethods int
-		for i, length := 0, methods.Len(); i < length; i++ {
-			methodDesc := methods.Get(i)
-			if transcoder.methods[methodPath(methodDesc)].httpRule != nil {
-				numSupportedMethods++
-			}
-		}
-		if numSupportedMethods == 0 {
-			return nil, fmt.Errorf("service %s only supports REST target protocol but has no methods with HTTP rules", svcDesc.FullName())
-		}
-	}
```

Comment:
We still need a check here -- if the service only has client and/or bidi stream methods that do _not_ use `google.api.HttpBody`, then we will not have any REST routes for it.

Reply:
We were missing checks to ensure a non http body couldn't be a target. I've moved to checking at runtime that the stream can be handled by the protocol. Any method can bind, but non http body streams will fail with an unsupported error. We may be able to relax this restriction in future versions. This feels more consistent than special casing REST rules.

Reply:
> This feels more consistent than special casing REST rules.

REST **is** a special case because it can't handle all kinds of methods whereas the other protocols can.

Failing at runtime is a terrible developer experience. If we know the config cannot work, we should fail fast.

Reply:
The config is not always controllable. For example these [stream proto files](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/examples/internal/proto/examplepb/stream.proto) are valid rules but we don't support the behaviour. I think we should still be able to register this service as this config may be valid behaviour in the future. I also don't see the terrible developer experience. For example if you register a service with one non-streaming and the other streaming endpoint the current behaviour will register the one and silently discard the other method. So now if you test the endpoints one will work and the other will return a 404. For this case I think it's much clearer to the user to fail with an explicit error that this streaming behaviour is not currently supported. The routes are valid but return an error.

Reply:
> For example if you register a service with one non-streaming and the other streaming endpoint the current behaviour will register the one and silently discard the other method.

Sure, that situation would not be an error. What I said was "if the service only has client and/or bidi stream methods that do not use google.api.HttpBody, then we will not have any REST routes for it." What I am saying is that it should be an error if there are **zero** valid/supported routes. That is also what this existing check was looking for.

> I also don't see the terrible developer experience.

If I configure a transcoder to _always_ translate to REST, and yet the transcoder will _always_ 404 because there are zero supported routes, that would be frustrating as a user to have that deploy and startup successfully and then not work at all. It is obvious that a transcoder with zero routes is not their intent. Failing fast means they can trivially check that the server is correctly configured in CI and also prevent bad roll-outs.